File: js/includes/cloud/availableServices.js
#!/usr/bin/env node

var request = require('request');
var path = require('path');

var __cloudDir = path.join(__dirname);

var RAPPCloud = require(path.join(__cloudDir, 'RAPPCloud.js'));

/**
 * @fileOverview Prototype the RAPPCloud Service Method.
 * 
 * @class availableServices
 * @memberof RAPPCloud
 * @description Asynchronous Service which will request the available_services cloud service for an Input
 * @version 1
 * @author Lazaros Penteridis <lp@ortelio.co.uk>
 * @param callback is the function that will receive the result
 */
RAPPCloud.prototype.availableServices = function ( callback )
{
	var cloud = this;
    var _delegate = callback;
	
	request.post({
        headers: {
//			'Authorization' : 'Basic cmFwcGRldjpyYXBwZGV2',
			'Content-Type' : 'application/x-www-form-urlencoded',
			'Connection' : 'close'
			},
        url: cloud.cloud_url + '/hop/available_services/ ',
    },
    function ( error, response, json ) 
    {
        if ( !error && response.statusCode == 200)
            handle_reply( json );
        else if ( error )
            error_handler ( error );
        else if ( response.statusCode != 200 )
            console.log ( 'Error: ' + response.statusCode );
    });
	
	function handle_reply( json )
    {
        var json_obj;
        try {
            json_obj = JSON.parse(json);
            // JSON reply is: { services: [], error: '' }
    
            if(json_obj.error){  // Check for Errors returned by the api.rapp.cloud
                    console.log('availableServices JSON error: ' + json_obj.error);
            }
            if (json_obj.services){
                    _delegate(json_obj.services);
            }
        } catch (e) {
            return console.error(e);
        }
    }
    
    function error_handler( error ) {
        return console.error(error);
    }  
};

/// Export
module.exports = RAPPCloud.availableServices;

File: cpp/includes/cloud/available_services/available_services.cpp
#include "available_services.hpp"

namespace rapp {
namespace cloud {

available_services::available_services(std::function<void(std::vector<service>)> callback)
: http_request("GET /hop/available_services HTTP/1.1\r\n"),
  delegate_(callback)
{
    assert(callback);
}

void available_services::deserialise(std::string json) 
{
    if (json.empty()) {
        throw std::runtime_error("empty json reply");
    }

    std::vector<service> services;
    auto json_f = json::parse(json);
    try {
        for (auto it_s : json_f["services"]) {
            services.push_back(std::make_pair(misc::get_json_value<std::string>("name", it_s), 
                                              misc::get_json_value<std::string>("url", it_s)));
        } 
        std::string error = misc::get_json_value<std::string>("error", json_f);
        if (!error.empty()) {
             std::cerr << "error JSON: " << error << std::endl; 
        }
    }
    catch (std::exception & e) {
        std::cerr << "Exception " << e.what() << std::endl;
    }
    
    delegate_(services);
}

}
}
